### PR TITLE
refactor: Change execution engine from drun to PocketIC. 

### PR DESCRIPTION
**Pull request by michael-weigelt**
This MR changes the execution engine from drun to PocketIC.

### CODE REVIEW

File: canbench-bin/src/lib.rs [hunk 1]
```diff
@@ -90,39 +91,41 @@ fn canbench_dir() -> PathBuf {
...
 }
 
-// Downloads drun if it's not already downloaded.
-fn maybe_download_drun(verbose: bool) {
-    const DRUN_LINUX_SHA: &str = "182b800a7979e1e3e516e54e4b9980e5407ced7464c0b3aec9ff7af6e9e69a1b";
-    const DRUN_MAC_SHA: &str = "8e0d0758d5a5c6f367e2c374dc7eae0106c7f46a3457f81018af6d5159d2dad4";
+// Downloads PocketIC if it's not already downloaded.
+fn maybe_download_pocket_ic(verbose: bool) {
+    const POCKET_IC_LINUX_SHA: &str =
+        "15feab44ebdc7b4b0157edd1a37ca07b8286262e7798abd531a5747b8fa8ea82";
```
Comment (michael-weigelt) [inline canbench-bin/src/lib.rs:102]: TODO: Both hashes need to be changed with the next PocketIC release.

File: canbench-bin/src/lib.rs [hunk 2]
```diff
@@ -90,39 +91,41 @@ fn canbench_dir() -> PathBuf {
...
+    // The expected version of PocketIC isn't present. Download it.
+    download_pocket_ic(verbose);
 }
 
-fn download_drun(verbose: bool) {
-    const DRUN_URL_PREFIX: &str =
-        "https://github.com/dfinity/ic/releases/download/release-2024-01-25_14-09/drun-x86_64-";
+fn download_pocket_ic(verbose: bool) {
+    const POCKET_IC_URL_PREFIX: &str =
+        "https://download.dfinity.systems/ic/5bcbdc7848400491ca9b738f3652a22094f21a3b/binaries/x86_64-";
```
Comment (michael-weigelt) [inline canbench-bin/src/lib.rs:128]: TODO: This path should be changed when the next version of PocketIC is released. 
Something like `https://github.com/dfinity/pocketic/releases/download/3.0.1/pocket-ic-x86_64-linux.gz`

File: Cargo.toml [hunk 3]
```diff
@@ -26,3 +26,4 @@ sha256 = "1.5.0"
 syn = "1.0"
 tempfile = "3.3.0"
 wasmparser = "0.119.0"
+pocket-ic = { path = "/home/mwe/dup/ic/packages/pocket-ic" }
```
Comment (michael-weigelt) [inline Cargo.toml:29]: TODO: Use next PocketIC version. The current one does not support large instruction limits.

File: canbench-bin/src/lib.rs [hunk 4]
```diff
@@ -139,68 +142,67 @@ fn download_drun(verbose: bool) {
...
+        .with_benchmarking_application_subnet()
+        .build();
+    let can_id = pic.create_canister();
+    pic.add_cycles(can_id, 1_000_000_000_000_000);
+    pic.install_canister(
+        can_id,
+        std::fs::read(canister_wasm_path).unwrap(),
+        vec![],
+        None,
+    );
```
[the remark this answers is not in the document]
Comment (michael-weigelt) [reply canbench-bin/src/lib.rs:180]: I agree.

**Comment by ielashi** [comment]
@michael-weigelt What's the status of this work?

**Comment by michael-weigelt** [comment]
@ielashi almost done.. we have released a new version of PocketIC and I am bumping the dependency

File: canbench-bin/src/lib.rs [hunk 5]
```diff
@@ -90,40 +91,41 @@ fn canbench_dir() -> PathBuf {
...
+    download_pocket_ic(verbose);
 }
 
-fn download_drun(verbose: bool) {
-    const DRUN_URL_PREFIX: &str =
-        "https://github.com/dfinity/ic/releases/download/release-2024-01-25_14-09/drun-x86_64-";
-
+fn download_pocket_ic(verbose: bool) {
+    const POCKET_IC_URL_PREFIX: &str =
+        "https://github.com/dfinity/pocketic/releases/download/4.0.0/pocket-ic-x86_64-";
```
[the remark this answers is not in the document]
Comment (mraszyk) [reply canbench-bin/src/lib.rs:128]: The 3.1 dependency is for the library which has separate version from the server.

File: canbench-bin/src/lib.rs [hunk 6]
```diff
@@ -139,68 +141,66 @@ fn download_drun(verbose: bool) {
...
-            eprintln!(
-                "Error executing benchmark {}. Error:\n{}",
-                bench_fn, output_str
-            );
+    // PocketIC is used for running the benchmark.
+    // Set the appropriate ENV variables
+    std::env::set_var("POCKET_IC_BIN", pocket_ic_path());
+    std::env::set_var("POCKET_IC_MUTE_SERVER", "1");
+    let pic = PocketIcBuilder::new()
+        .with_benchmarking_system_subnet()
```
[the remark this answers is not in the document]
Comment (mraszyk) [reply canbench-bin/src/lib.rs:169]: yes

[the remark this answers is not in the document]
Comment (mraszyk) [reply canbench-bin/src/lib.rs:169]: Here's the actual instruction limit: https://github.com/dfinity/ic/blob/1b76908a3c47876a58e2500ec1b46c90d9d5cbd8/rs/pocket_ic_server/src/pocket_ic.rs#L163

**Review by ielashi — APPROVED** [review]
LGTM, thanks for putting this PR together!

**Pull request merged**